packages/graphiql-react/src/editor/operations-editor.ts

```diff
@@ -0,0 +1,150 @@
+import { DOC_EXPLORER_PLUGIN, usePluginContext } from '../plugin';
+import { useSchemaContext } from '../schema';
+import { useStorageContext } from '../storage';
+import debounce from '../utility/debounce';
+import { DEFAULT_EDITOR_THEME, DEFAULT_KEY_MAP } from './common';
+import { useEditorContext } from './context';
+import {
+  useCopyQuery,
+  UseCopyQueryArgs,
+  useMergeQuery,
+  usePrettifyEditors,
+} from './hooks';
+import { WriteableEditorProps } from './types';
+import { KeyCode, KeyMod } from 'monaco-editor';
+import { MONACO_GRAPHQL_API, OPERATIONS_MODEL } from '@/constants';
+import { createEditor } from '@/create-editor';
+
+export type UseQueryEditorArgs = WriteableEditorProps &
+  Pick<UseCopyQueryArgs, 'onCopyQuery'> & {
+    /**
+     * Invoked when a reference to the GraphQL schema (type or field) is clicked
+     * as part of the editor or one of its tooltips.
+     * @param reference The reference that has been clicked.
+     */
+    onClickReference?(reference: SchemaReference): void;
+    /**
+     * Invoked when the contents of the query editor change.
+     * @param value The new contents of the editor.
+     * @param documentAST The editor contents parsed into a GraphQL document.
+     */
+    onEdit?(value: string, documentAST?: DocumentNode): void;
+  };
+
+export function useQueryEditor(
+  {
+    editorTheme = DEFAULT_EDITOR_THEME,
+    keyMap = DEFAULT_KEY_MAP,
+    onClickReference,
+    onCopyQuery,
+    onEdit,
+    readOnly = false,
+  }: UseQueryEditorArgs = {},
+  caller?: Function,
+) {
+  const { schema } = useSchemaContext({
+    nonNull: true,
+    caller: caller || useQueryEditor,
+  });
+  const {
+    externalFragments,
+    initialQuery,
+    queryEditor,
+    setOperationName,
+    setQueryEditor,
+    validationRules,
+    variableEditor,
+    updateActiveTabValues,
+  } = useEditorContext({
+    nonNull: true,
+    caller: caller || useQueryEditor,
+  });
+  const executionContext = useExecutionContext();
+  const storage = useStorageContext();
+  const explorer = useExplorerContext();
+  const plugin = usePluginContext();
+  const copy = useCopyQuery({ caller: caller || useQueryEditor, onCopyQuery });
+  const merge = useMergeQuery({ caller: caller || useQueryEditor });
+  const prettify = usePrettifyEditors({ caller: caller || useQueryEditor });
+  const ref = useRef<HTMLDivElement>(null);
+  const { run } = executionContext!;
+
+  useEffect(() => {
+    setQueryEditor(createEditor('operations', ref.current!));
+    OPERATIONS_MODEL.onDidChangeContent(
+      debounce(100, () => {
+        const value = OPERATIONS_MODEL.getValue();
+        storage?.set(STORAGE_KEY_QUERY, value);
+        updateActiveTabValues({
+          query: value,
+          operationName: /* operationFacts?.operationName ?? */ null,
+        });
+      }),
+    );
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- only on mount
+  }, []);
+
+  useEffect(() => {
```

Comment:
this should run on load, just once and ideally right after you have the editor created. currently, results in this:
<img width="925" alt="Screenshot 2023-06-20 at 6 23 49 PM" src="https://github.com/graphql/graphiql/assets/963528/6e7b99d2-a175-4f0d-a91e-a3006fe6d89e">

Reply:
we should expose a hook that allows adding commands without lifecycle issues. this can be the replacement for extraKeys prop which should be removed